registry/registry.go

```diff
@@ -172,65 +58,49 @@ func (c *client) GetRepository(repository string) (_ []flux.ImageDescription, er
 	// `library/nats`. We need that to fetch the tags etc. However, we
 	// want the results to use the *actual* name of the images to be
 	// as supplied, e.g., `nats`.
-	return c.tagsToRepository(cancel, client, hostlessImageName, repository, tags)
+	return c.tagsToRepository(r, img, tags)
 }
 
-func (c *client) lookupImage(client backend, lookupName, imageName, tag string) (flux.ImageDescription, error) {
-	// Minor cheat: this will give the correct result even if the
-	// imageName includes a host
-	id := flux.MakeImageID("", imageName, tag)
-	img := flux.ImageDescription{ID: id}
-
-	start := time.Now()
-	history, err := client.Manifest(lookupName, tag)
-	c.Metrics.RequestDuration.With(
-		LabelRepository, imageName,
-		LabelRequestKind, RequestKindMetadata,
-		fluxmetrics.LabelSuccess, strconv.FormatBool(err == nil),
-	).Observe(time.Since(start).Seconds())
+// Get a single Image from the registry if it exists
+func (c *registry) GetImage(img Repository, tag string) (_ Image, err error) {
+	r, err := c.newRemote(img)
 	if err != nil {
-		return img, err
-	}
-	// the manifest includes some v1-backwards-compatibility data,
-	// oddly called "History", which are layer metadata as JSON
-	// strings; these appear most-recent (i.e., topmost layer) first,
-	// so happily we can just decode the first entry to get a created
-	// time.
-	type v1image struct {
-		Created time.Time `json:"created"`
-	}
-	var topmost v1image
-	if err = json.Unmarshal([]byte(history[0].V1Compatibility), &topmost); err == nil {
-		if !topmost.Created.IsZero() {
-			img.CreatedAt = &topmost.Created
-		}
+		return
 	}
+	return r.Manifest(img, tag)
+}
 
-	return img, err
+func (c *registry) newRemote(img Repository) (remote Remote, err error) {
+	remote, err = c.factory.CreateFor(img.Host())
+	if err != nil {
+		return
+	}
+	remote = NewInstrumentedRemote(c.Metrics)(remote)
+	return
 }
 
-func (c *client) tagsToRepository(cancel func(), client backend, lookupName, imageName string, tags []string) ([]flux.ImageDescription, error) {
+func (c *registry) tagsToRepository(remote Remote, repository Repository, tags []string) ([]Image, error) {
 	// one way or another, we'll be finishing all requests
-	defer cancel()
+	defer remote.Cancel()
 
 	type result struct {
-		image flux.ImageDescription
+		image Image
 		err   error
 	}
 
 	fetched := make(chan result, len(tags))
 
 	for _, tag := range tags {
 		go func(t string) {
-			img, err := c.lookupImage(client, lookupName, imageName, t)
+			i, err := remote.Manifest(repository, t)
 			if err != nil {
 				c.Logger.Log("registry-metadata-err", err)
 			}
-			fetched <- result{img, err}
+			fetched <- result{i, err}
 		}(tag)
 	}
 
-	images := make([]flux.ImageDescription, cap(fetched))
+	images := make([]Image, cap(fetched))
```

Comment:
Just to elaborate. ImageDescription is bad because there isn't an entity called ImageDescription. Image HAS A Description, yes, but ImageDescription isn't a thing. And in this case, it's got nothing to do with a description. Its a "created at" time.

It should be something like:
```Go
type Image struct {
    id string
    createdAt time.Time
}
```
I would have made this change, but as mentioned in another comment, ImageID is too coupled to the rest of the codebase and is too ill-defined to state it's use (it's just a string).